src/modules/content-s3/s3.c

```diff
@@ -0,0 +1,268 @@
+    ctx->count += size;
+
+    return size;
+}
+
+/* Gets the object from s3 storing it in in 'data->datap' as well
+ * as updating 'data->sizep' to hold the size of the data read.
+ */
+static S3Status get_object_cb (int buff_size, const char *buff, void *data)
+{
+    void *tmp = NULL;
+    struct cb_data *ctx = data;
+
+    if (! (tmp = realloc (ctx->data, buff_size + ctx->size)))
+        return S3StatusOutOfMemory;
+    ctx->data = tmp;
+
+    memcpy (ctx->data + ctx->size, buff, buff_size);
+    ctx->size += buff_size;
+
+    return S3StatusOK;
+}
+
+int s3_init (struct s3_config *cfg, const char **errstr)
+{
+    S3Status status = S3_initialize ("s3", S3_INIT_ALL, cfg->hostname);
+
+    if (status != S3StatusOK) {
+        errno = ECONNREFUSED;
+        if (errstr)
+            *errstr = S3_get_status_name (status);
+
+        return -1;
+    }
+    
+    return 0;
+}
+
+void s3_cleanup ()
+{
+    S3_deinitialize ();
+}
+
+int s3_bucket_create (struct s3_config *cfg, const char **errstr)
+{
+    int retries = cfg->retries;
+    S3Status status = S3_validate_bucket_name (cfg->bucket, uri_style);
+
+    if (status != S3StatusOK) {
+        errno = ECONNREFUSED;
+        if (errstr)
+            *errstr = S3_get_status_name (status);
+
+        return -1;
+    }
+
+    S3ResponseHandler resp_hndl = {
+        .propertiesCallback = &response_props_cb,
+        .completeCallback = &response_complete_cb
+    };
+
+    struct cb_data ctx = {
+        0,
+        NULL,
+        0,
+        status,
+    };
+
+    do {
+        S3_create_bucket (protocol, cfg->access_key, cfg->secret_key, 0, cfg->bucket, S3CannedAclPrivate, 0, 0, &resp_hndl, &ctx);
+        retries--;
+    } while (S3_status_is_retryable (ctx.status) && retries > 0);
+
+    if (ctx.status != S3StatusOK && ctx.status != S3StatusErrorBucketAlreadyOwnedByYou) {
+        errno = EREMOTEIO;
+        if (errstr)
+            *errstr = S3_get_status_name (status);
+
+        return -1;
+    }
+
+    return 0;
+}
+
+int s3_put (struct s3_config *cfg, const char *key, const void *data, size_t size, const char **errstr)
+{
+    int retries = cfg->retries;
+    S3Status status = S3StatusOK;
+    
+    S3ResponseHandler resp_hndl = {
+        .propertiesCallback = &response_props_cb,
+        .completeCallback = &response_complete_cb
+    };
+
+    S3BucketContext bucket_ctx = {
+        0,
+        cfg->bucket,
+        protocol,
+        uri_style,
+        cfg->access_key,
+        cfg->secret_key
+    };
+
+    S3PutObjectHandler put_obj_hndl ={
+        .responseHandler = resp_hndl,
+        .putObjectDataCallback = &put_object_cb
+    };
+
+    if (strlen (key) == 0 || strchr (key, '/') || !strcmp (key, "..") || !strcmp (key, ".")) {
+        errno = EINVAL;
+        if (errstr)
+            *errstr = "invalid key";
+
+        return -1;
+    }
+
+    struct cb_data ctx = {
+        size,
+        (void *) data,
+        0,
+        status
```

Comment:
use tagged initializers

also, the cast to `(void *)` is unnecessary - the structure member is defined as a `void *`.

Reply:
If I remove the cast to `(void *)` I get this error from make: `error: initialization discards ‘const’ qualifier from pointer target type` because the data is passed in as a `const void *`

Reply:
Oops!  My bad.  You are right - keep the cast!